File: adventofcode_intcode/src/main/kotlin/day22/day22.kt
package day22

import java.io.File
import java.math.BigInteger

private fun mod(a: BigInteger, m: BigInteger): BigInteger = (a % m + m) % m
private fun modInv(a: BigInteger, m: BigInteger): BigInteger = a.modPow(m - BigInteger.TWO, m)

private fun parse(input: List<String>, m: BigInteger, n: BigInteger): Pair<BigInteger, BigInteger> {
    var a = BigInteger.ONE
    var b = BigInteger.ZERO

    for (line in input) {
        when {
            "into new stack" in line -> {
                a = -a
                b = m - b - BigInteger.ONE
            }
            "cut" in line -> {
                val i = BigInteger(line.split(" ").last())
                b = (m + b - i) % m
            }
            "increment" in line -> {
                val i = BigInteger(line.split(" ").last())
                a = (a * i) % m
                b = (b * i) % m
            }
            else -> throw Exception("Invalid instruction")
        }
    }

    val resA = a.modPow(n, m)
    val resB = (b * (a.modPow(n, m) - BigInteger.ONE) * modInv(a - BigInteger.ONE, m)) % m

    return resA to resB
}

fun evaluatorOne(input: List<String>): BigInteger {
    val m = BigInteger.valueOf(10007)
    val iter = BigInteger.ONE
    val (a, b) = parse(input, m, iter)
    return mod(a * BigInteger.valueOf(2019) + b, m)
}

fun evaluatorTwo(input: List<String>): BigInteger {
    val m = BigInteger.valueOf(119315717514047)
    val iter = BigInteger.valueOf(101741582076661)
    val (a, b) = parse(input, m, iter)
    return mod(modInv(a, m) * (BigInteger.valueOf(2020) - b), m)
}

fun readLinesFromFile(filePath: String): Result<List<String>> =
    runCatching { File(filePath).readLines() }

fun main() {
    readLinesFromFile("src/main/resources/day22.txt")
        .onSuccess {
            println("Part One: ${evaluatorOne(it)}")
            println("Part Two: ${evaluatorTwo(it)}")
        }
        .onFailure {
            println("Error reading file: ${it.message}")
        }
}

File: 2019/src/main/scala/day22/day22.scala
package day22

import scala.util.{Try, Success, Failure, Using}
import scala.io.Source

case class Technique[M <: Long](a: BigInt, c: BigInt)(using v: ValueOf[M]) {
    private val m = BigInt(v.value)

    infix def compose(other: Technique[M]): Technique[M] = {
        val a = (this.a * other.a) % m
        val c = (this.c * other.a + other.c) % m
        Technique[M](a, c)
    }

    def inverse(): Technique[M] = {
        val a = this.a.modInverse(m)
        val c = m - (a * this.c) % m
        Technique[M](a, c)
    }

    def power(e: BigInt): Technique[M] = {
        val a = this.a.modPow(e, m)
        val c = (((a - 1) * (this.a - 1).modInverse(m) % m) * this.c) % m
        Technique[M](a, c)
    }

    def shuffle(index: BigInt): BigInt = {
        (this.a * index + this.c) % m
    }
}

def deck[M <: Long](input: List[String])(using v: ValueOf[M]): Technique[M] = {
    val m = BigInt(v.value)

    val techniques = input.collect {
        case s"cut $num" => {
            val n = BigInt(num)
            val c = (m - n % m) % m
            Technique[M](1, c)
        }
        case "deal into new stack" => {
            Technique[M](m - 1, m - 1)
        }
        case s"deal with increment $num" => {
            val n = BigInt(num)
            val a = (m + n % m) % m;
            Technique[M](a, 0)
        }
    }

    return techniques.reduce(_ compose _)
}

def evaluatorOne(input: List[String]): BigInt = {
    return deck[10007L](input).shuffle(2019)
}

def evaluatorTwo(input: List[String]): BigInt = {
    return deck[119315717514047L](input).inverse().power(101741582076661L).shuffle(2020)
}

def readLinesFromFile(filePath: String): Try[List[String]] =
    Using(Source.fromResource(filePath))(_.getLines().toList)

def hello(): Unit = {
    readLinesFromFile("day22.txt") match {
        case Success(lines) => {
            println(s"Part One: ${evaluatorOne(lines)}")
            println(s"Part Two: ${evaluatorTwo(lines)}")
        }
        case Failure(exception) => {
            println(s"Error reading file: ${exception.getMessage}")
        }
    }
}